File: processor/src/main/java/com/lsxiao/apllo/processor/ApolloDescriptor.java
package com.lsxiao.apllo.processor;

import com.apollo.core.entity.SchedulerProvider;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.ExecutableElement;

import io.reactivex.BackpressureStrategy;

/**
 * write with Apollo
 * author:lsxiao
 * date:2017-04-22 02:19
 * github:https://github.com/lsxiao
 * zhihu:https://zhihu.com/people/lsxiao
 */

public class ApolloDescriptor {
    private ExecutableElement mMethodElement;
    private boolean mIsSticky = false;
    private List<String> mTags = new ArrayList<>();
    private SchedulerProvider.Tag mSubscribeOn = SchedulerProvider.Tag.IO;
    private SchedulerProvider.Tag mObserveOn = SchedulerProvider.Tag.MAIN;
    private BackpressureStrategy mBackpressureStrategy;

    private ApolloDescriptor(ExecutableElement methodElement) {
        mMethodElement = methodElement;
    }

    public static ApolloDescriptor newInstance(ExecutableElement methodElement) {
        return new ApolloDescriptor(methodElement);
    }

    public ApolloDescriptor sticky(boolean val) {
        mIsSticky = val;
        return this;
    }

    public ApolloDescriptor tags(List<String> val) {
        mTags = val;
        return this;
    }

    public ApolloDescriptor subscribe(SchedulerProvider.Tag val) {
        mSubscribeOn = val;
        return this;
    }

    public ApolloDescriptor observe(SchedulerProvider.Tag val) {
        mObserveOn = val;
        return this;
    }

    public ApolloDescriptor backpressure(BackpressureStrategy val) {
        mBackpressureStrategy = val;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ApolloDescriptor that = (ApolloDescriptor) o;

        return mMethodElement.equals(that.mMethodElement);

    }

    @Override
    public int hashCode() {
        return mMethodElement.hashCode();
    }

    public ExecutableElement getMethodElement() {
        return mMethodElement;
    }

    public boolean isSticky() {
        return mIsSticky;
    }

    public List<String> getTags() {
        return mTags;
    }

    public SchedulerProvider.Tag getSubscribeOn() {
        return mSubscribeOn;
    }

    public SchedulerProvider.Tag getObserveOn() {
        return mObserveOn;
    }

    public BackpressureStrategy getBackpressureStrategy() {
        return mBackpressureStrategy;
    }
}

File: processor/src/main/java/com/lsxiao/apllo/processor/ApolloDescriptor.kt
package com.lsxiao.apllo.processor

import com.lsxiao.apollo.core.entity.SchedulerProvider
import io.reactivex.BackpressureStrategy
import java.util.*
import javax.lang.model.element.ExecutableElement

/**
 * write with Apollo
 * author:lsxiao
 * date:2017-04-22 02:19
 * github:https://github.com/lsxiao
 * zhihu:https://zhihu.com/people/lsxiao
 */

class ApolloDescriptor private constructor(val methodElement: ExecutableElement) {
    companion object {
        fun newInstance(methodElement: ExecutableElement): ApolloDescriptor {
            return ApolloDescriptor(methodElement)
        }
    }

    var isSticky = false
    var stickyAutoRemove = true
    var take = 0
    var tags: List<String> = ArrayList()
    var subscribeOn = SchedulerProvider.Tag.IO
    var observeOn = SchedulerProvider.Tag.MAIN
    var backpressureStrategy: BackpressureStrategy? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as ApolloDescriptor?

        return methodElement == that!!.methodElement

    }

    override fun hashCode(): Int {
        return methodElement.hashCode()
    }

}
